service/matching/matchingEngine.go

```diff
@@ -1084,6 +1109,80 @@ func (e *matchingEngineImpl) emitForwardedSourceStats(
 	}
 }
 
+func (e *matchingEngineImpl) redirectToVersionedQueueForPoll(
+	ctx context.Context,
+	taskQueue *taskQueueID,
+	workerVersionCapabilities *commonpb.WorkerVersionCapabilities,
+	kind enumspb.TaskQueueKind,
+) (*taskQueueID, error) {
+	// sticky queues are unversioned
+	if kind == enumspb.TASK_QUEUE_KIND_STICKY {
+		return taskQueue, nil
+	}
+	unversionedTQM, err := e.getTaskQueueManager(ctx, taskQueue, kind, true)
+	if err != nil {
+		return nil, err
+	}
+	userData, err := unversionedTQM.GetUserData(ctx)
```

Comment:
I see you mentioned in the PR description that this isn't quite right yet so I'm not blocking this PR but just noting that we don't want to get user data on every request, we need to throttle this.

Reply:
GetUserData just returns what's cached in memory.

Throttling is fine, but that should be handled at the persistence level.

Reply:
Am I reading this wrong?
`GetUserData` returns what's cached or reads through.

Reply:
Right, it returns what's cached in memory. That's what we want, no? Of course we can't go to the db on every task

Reply:
My point was that if there's a cache miss, we'll go to the DB on every task.
It's fine if the throttling happens at a different level but read for every cache miss is not what we want.
As long as we're tracking, this isn't blocking _this_ PR.

Reply:
I think you're mixing two separate things here: loading the version data (what this is doing) and possibly refreshing it if we have some data but we're looking for a specific build id that isn't present ("read-through"). the read-through part isn't implemented yet, so let's talk about that later.

as for the loading, it can happen from the db (for the root) or from the root. for either one, once we do a read successfully, whether we have data or not, then we won't do it again (ignore the read-through for now). I guess it's possible for a few calls to go through while the first is in flight, so that's something to pay attention to (later). but I don't see any way it can "go to the DB on every task"

well, I suppose unless the db read fails continually. anyway, I have an idea for how to handle this (in future PRs).